src/kv/LockResolver.cc

```diff
@@ -253,11 +271,179 @@ void LockResolver::resolvePessimisticLock(Backoffer & bo, LockPtr lock, std::uno
+
+        auto [regions, region_id] = cluster->region_cache->groupKeysByRegion(bo, keys);
+        (void)region_id;
+        for (auto & [region_id, keys] : regions)
+        {
+            resolveRegionLocks(bo, lock, region_id, keys, status);
+        }
+        return;
+    }
+    if (response == nullptr)
+    {
+        throw Exception("Response body missing.", ErrorCodes::UnknownError);
+    }
+}
+
+AsyncResolveDataPtr LockResolver::checkAllSecondaries(Backoffer & bo, LockPtr lock, TxnStatus & status)
+{
+    std::vector<std::string> secondaries;
+    secondaries.reserve(status.primary_lock->secondaries_size());
+    for (size_t i = 0; i < status.primary_lock->secondaries_size(); i++)
+    {
+        secondaries.push_back(status.primary_lock->secondaries(i));
+    }
+
+    auto [regions, region_id] = cluster->region_cache->groupKeysByRegion(bo, secondaries);
+    (void)region_id;
+    auto shared_data = std::make_shared<AsyncResolveData>(status.primary_lock->min_commit_ts(), false);
+    std::vector<std::thread> threads;
+    std::atomic_int8_t errors{0};
+    for (auto & pair : regions)
+    {
+        auto & region_id = pair.first;
+        auto & keys = pair.second;
+        threads.emplace_back([&]() {
+            try
+            {
+                checkSecondaries(bo, lock->txn_id, keys, region_id, shared_data);
+            }
+            catch (Exception & e)
+            {
+                errors.fetch_add(1);
+                log->error("checkSecondaries error: " + e.displayText());
+            }
+        });
+    }
+
+    for (auto & t : threads)
+    {
+        t.join();
+    }
+
+    if (errors.load() > 0)
+    {
+        throw Exception("resolveAsyncLock failed", ErrorCodes::UnknownError);
+    }
+
+    return shared_data;
+}
+
+void LockResolver::checkSecondaries(
+    Backoffer & bo, uint64_t txn_id, std::vector<std::string> & cur_keys, RegionVerID cur_region_id, AsyncResolveDataPtr shared_data)
+{
+    auto check_request = std::make_shared<::kvrpcpb::CheckSecondaryLocksRequest>();
+    for (auto & key : cur_keys)
+    {
+        auto * k = check_request->add_keys();
+        *k = key;
+    }
+    check_request->set_start_version(txn_id);
+
+    RegionClient client(cluster, cur_region_id);
+    std::shared_ptr<::kvrpcpb::CheckSecondaryLocksResponse> response{};
+    try
+    {
+        response = client.sendReqToRegion(bo, check_request);
+    }
+    catch (Exception & e)
+    {
+        bo.backoff(boRegionMiss, e);
+        auto [regions, region_id] = cluster->region_cache->groupKeysByRegion(bo, cur_keys);
+        (void)region_id;
+
+        for (auto & [region_id, keys] : regions)
+        {
+            checkSecondaries(bo, txn_id, keys, region_id, shared_data);
+        }
+        return;
+    }
+    if (response == nullptr)
+    {
+        throw Exception("Response body missing.", ErrorCodes::UnknownError);
+    }
+
+    shared_data->addKeys(response, cur_keys.size(), txn_id, response->commit_ts());
```

Comment:
commit_ts can be derived from response, why do we pass it ?

Reply:
It's because they just pass a slice in TiDB instead of the response. I'll refine it.